# Assignment 5
###### By Ian, Eric, Johnny

This assignment is meant to test the confidence intervals of different decision trees. Here I test 5 trees, starting with high depth restriction and 'loosening' the restriction by 1 level each time. The output shows the confidence interval corresponding to each tree.

In [1]:
# Import necessary modules
import pandas as pd
from sklearn import tree
from bootstrap import bootstrap

# Was getting several 'FutureWarning's so I just went ahead and silenced them 
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

# Classify the trees
t1 = tree.DecisionTreeClassifier(criterion='entropy', max_depth=2)
t2 = tree.DecisionTreeClassifier(criterion='entropy', max_depth=3)
t3 = tree.DecisionTreeClassifier(criterion='entropy', max_depth=4)
t4 = tree.DecisionTreeClassifier(criterion='entropy', max_depth=5)
t5 = tree.DecisionTreeClassifier(criterion='entropy', max_depth=None)


print("Caesarian\n=====================================================================")

# Read csv data to dataframe
df = pd.read_csv("caesarian.csv")

# Show the confidence intervals
print("Confidence interval max_depth = 2: {}".format(bootstrap(t1,df,'Caesarian')))
print("Confidence interval max_depth = 3: {}".format(bootstrap(t2,df,'Caesarian')))
print("Confidence interval max_depth = 4: {}".format(bootstrap(t3,df,'Caesarian')))
print("Confidence interval max_depth = 5: {}".format(bootstrap(t4,df,'Caesarian')))
print("Confidence interval max_depth = None: {}".format(bootstrap(t5,df,'Caesarian')))

Caesarian
Confidence interval max_depth = 2: (0.43593750000000003, 0.875)
Confidence interval max_depth = 3: (0.375, 0.9375)
Confidence interval max_depth = 4: (0.4375, 0.9375)
Confidence interval max_depth = 5: (0.49843750000000003, 0.9375)
Confidence interval max_depth = None: (0.5625, 1.0)


## Report
Above we see that the best tree provides clearly the best confidence interval, being between above 56% to 100% accuracy. This means that no matter what the accuracy of the model is most likely correct (since >50% gives positive odds). As the model is restricted, however, we can see the interval slip below 50% which would result in prediction to inevetably be wrong on some occasions.

## Team Assessment
Overall the team worked well, helping eachother with issues they were having. Our body of work is all different because we worked on separate machines, but there was good communication between everyone in helping us all get a final product. Each teammate did equal amounts of work and deserves equal credit.